packages/contracts/contracts/plugin/psp/PluginSetupProcessor.sol

```diff
@@ -0,0 +1,703 @@
+            versionTag: _params.pluginSetupRef.versionTag,
+            data: _params.data,
+            plugin: plugin,
+            preparedDependency: preparedDependency
+        });
+
+        return (plugin, preparedDependency);
+    }
+
+    /// @notice Applies the permissions of a prepared installation to a DAO.
+    /// @param _dao The address of the installing DAO.
+    /// @param _params The struct containing the parameters for the `applyInstallation` function.
+    function applyInstallation(address _dao, ApplyInstall calldata _params)
+        external
+        canApply(_dao, APPLY_INSTALLATION_PERMISSION_ID)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.plugin);
+
+        PluginInformation storage pluginInformation = states[pluginId];
+
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            pHash(_params.permissions),
+            _params.helpersHash,
+            aHash(_params.actions),
+            bytes("")
+        );
+
+        // Allow calling `applyInstallation` only when
+        // plugin was uninstalled or never been installed before.
+        if (pluginInformation.currentSetupId != bytes32(0)) {
+            revert PluginAlreadyInstalled();
+        }
+
+        // If the plugin block number exceeds the setupId preparation block number,
+        // This means applyInstallation was already called on another setupId
+        // and all the rest setupIds should become idle or setupId is not prepared before.
+        if (pluginInformation.blockNumber >= pluginInformation.setupIds[setupId]) {
+            revert SetupNotEligible(setupId);
+        }
+
+        bytes32 newSetupId = _getSetupId(
+            _params.pluginSetupRef,
+            bytes32(0),
+            _params.helpersHash,
+            bytes32(0),
+            bytes("")
+        );
+
+        pluginInformation.currentSetupId = newSetupId;
+        pluginInformation.blockNumber = block.number;
+
+        _executeOnDAO(_dao, setupId, _params.permissions, _params.actions);
+
+        emit InstallationApplied({dao: _dao, plugin: _params.plugin, setupId: setupId});
+    }
+
+    /// @notice Prepares the update of an UUPS upgradeable plugin.
+    /// @param _dao The address of the installing DAO.
+    /// @param _params The struct containing the parameters for the `prepareUpdate` function.
+    /// @return initData The initialization data to be passed to upgradeable contracts when the update is applied
+    /// @return preparedDependency TOD:GIORGI
+    /// @dev The list of `_currentHelpers` has to be specified in the same order as they were returned from previous setups preparation steps (the latest `prepareInstallation` or `prepareUpdate` step that has happend) on which the update is prepared for
+    function prepareUpdate(address _dao, PrepareUpdate calldata _params)
+        external
+        returns (bytes memory initData, IPluginSetup.PreparedDependency memory preparedDependency)
+    {
+        if (
```

Comment:
are these checks correct?
the release has to be the same and the build to be bigger to continue the update.
shouldn't this check that it release is the same build has to be bigger

Reply:
fixed in https://github.com/aragon/core/pull/164/commits/047027d41007b101714e43727df616dc8910ca78